File: filter.js

function Volout(min, max, len, largestDimensions, res){
    this.min = min;
    this.max = max;
    this.len = len;
    this.ld = largestDimensions;
    this.ar = new Array(res*res*res);
    this.res = res;
    this.ressq = res * res;
    for(var i = 0; i < this.ar.length; i++) this.ar[i] = [0,0,0,0];
}

Volout.prototype.push = function(x,y,z, r,g,b){
    x -= this.min[0];
    y -= this.min[1];
    z -= this.min[2];
    
    x *= this.res / this.len[0];
    y *= this.res / this.len[1];
    z *= this.res / this.len[2];
    
    x = Math.floor(x);
    y = Math.floor(y);
    z = Math.floor(z);
    
    if(x < 0 || x >= this.res || y < 0 || y >= this.res || z < 0 || z >= this.res) return;
    
    var index = z * this.ressq + y * this.res + x;
    
    var c = this.ar[index][3];
    var scalar = c / (c+1);
    var scalar2 = 1 / (c+1);
    this.ar[index][0] = r * scalar + this.ar[index][0] * scalar2;
    this.ar[index][1] = g * scalar + this.ar[index][1] * scalar2;
    this.ar[index][2] = b * scalar + this.ar[index][2] * scalar2;
    
    this.ar[index][3]++;
    
};

Volout.prototype.write = function(fileName){
    var fs = require('fs');
    var out = fs.createWriteStream(fileName);
    var x,y,z;
    var has = false;
    for(z=0;z<this.res;z++){
        for(y=0;y<this.res;y++){
            for(x=0;x<this.res;x++){
                var index = z * this.ressq + y * this.res + x;
                var a = this.ar[index];
                if(a[3] > 0){
                    if(has) out.write(';');
                    var tmp = [x,y,z].concat(a.slice(0,3)).map(Math.floor.bind(Math));
                    out.write(tmp.join(','));
                    has= true;
                }
                
            }
        }
    }
    out.end();
};



function stats(fileName, cb){
    var LLine = require('./LLine.js');
    var min = [Number.MAX_VALUE,Number.MAX_VALUE,Number.MAX_VALUE];
    var max = [Number.NEGATIVE_INFINITY,Number.NEGATIVE_INFINITY,Number.NEGATIVE_INFINITY];
    
    LLine.read(fileName, function(line){
        var nums = line.split(' ').map(Number);
        if(nums.length != 6) return;
        for(var i = 0; i < 3; i++){
            if(nums[i] > max[i]) max[i] = nums[i];
            if(nums[i] < min[i]) min[i] = nums[i];
        }
        
        
    }, function(){
        var lengths = min.map((m, i) => max[i] - m);
        var ret = {
            min:min, max:max,
            len : lengths,
            largestDimension : Math.max.apply(Math, lengths)
        };
        cb(ret);
    });
}

var readFileName = 'myOutput.txt';

stats(readFileName, function(d){
    console.log('stats', d);
    var v = new Volout(d.min, d.max, d.len, d.largestDimension, 128);
    require('./LLine.js').read(readFileName, function(line){
        var nums = line.split(' ').map(Number);
        v.push(nums[0],nums[1],nums[2],nums[3],nums[4],nums[5]);
    }, function(){
        console.log('writing output');
        v.write('pc.txt');
    });
});

File: filter.py
import math

class Volout:
    def __init__(self, minIn, maxIn, res):
        self.min = minIn
        self.max = maxIn
        self.len = [self.max[i]-self.min[i] for i in range(3)]
        self.ld = max(self.len[0], self.len[1], self.len[2])
        self.ar = [[0,0,0,0] for i in range(res*res*res)]
        self.res = res
        self.ressq = res * res

    def push(self, x,y,z, r,g,b):
        x -= self.min[0]
        y -= self.min[1]
        z -= self.min[2]

        x *= self.res / self.len[0]
        y *= self.res / self.len[1]
        z *= self.res / self.len[2]

        x = int(math.floor(x))
        y = int(math.floor(y))
        z = int(math.floor(z))

        if x < 0 or x >= self.res or y < 0 or y >= self.res or z < 0 or z >= self.res:
            return

        index = z * self.ressq + y * self.res + x;

        c = self.ar[index][3]
        scalar = c / float(c+1)
        scalar2 = 1 / float(c+1)
        self.ar[index][0] = r * scalar + self.ar[index][0] * scalar2
        self.ar[index][1] = g * scalar + self.ar[index][1] * scalar2
        self.ar[index][2] = b * scalar + self.ar[index][2] * scalar2

        self.ar[index][3] += 1

    def write(self,fileName):
        out = open(fileName, 'w')
        has = False
        for z in range(self.res):
            for y in range(self.res):
                for x in range(self.res):
                    index = z * self.ressq + y * self.res + x
                    a = self.ar[index]
                    if a[3] > 0:
                        if has: out.write(';')
                        tmp = [x,y,z] + [int(b) for b in a[0:3]] #[x,y,z].concat(a.slice(0,3)).map(Math.floor.bind(Math));
                        out.write(','.join([str(i) for i in tmp]))
                        has = True
        out.close()
        
def readStats(fileName):
    min = None
    max = None
    file = open(fileName, 'r')
    for line in file:
        if line[0] != '#':
            nums = [float(i) for i in line.split(',')]
            if min != None:
                for j in range(3):
                    min[j] = min[j] if min[j] < nums[j] else nums[j]
                    max[j] = max[j] if max[j] > nums[j] else nums[j]
            else:
                min = nums[0:3]
                max = nums[0:3]
    file.close()
    return min, max

fn = 'multi_selection.csv'
mn, mx = readStats(fn)
vo = Volout(mn, mx, 256)
file = open(fn, 'r')
for line in file:
    if line[0] != '#':
        nums = [float(i) for i in line.split(',')]
        vo.push(nums[0], nums[1], nums[2], nums[3], nums[4], nums[5])
        #print nums
file.close()
vo.write('pc.txt')
    
    
